common/util/iterator.go

```diff
@@ -0,0 +1,165 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package util
+
+import (
+	"bytes"
+	"errors"
+	"sync"
+)
+
+var (
+	// ErrIteratorFinished indicates that Next was called on an iterator
+	// which has already reached the end of its input.
+	ErrIteratorFinished = errors.New("iterator has reached end")
+)
+
+type (
+	// Iterator is used to iterate over entities.
+	// Each entity is represented as a byte slice.
+	// Iterator will fetch pages using the provided GetFn.
+	// Pages will be fetched starting from provided minPage and continuing until provided maxPage.
+	// Iterator will skip over any pages with empty input and will skip over any empty elements within a page.
+	// Iterator is thread safe and makes deep copies of all in and out data.
+	// If Next returns an error all subsequent calls to Next will return the same error.
+	Iterator interface {
+		Next() ([]byte, error)
+		HasNext() bool
+	}
+
+	// GetFn fetches bytes for given page number. Returns error on failure.
+	GetFn func(int) ([]byte, error)
+
+	iterator struct {
+		sync.Mutex
+
+		currentPage int
+		page        [][]byte
+		pageIndex   int
+		nextResult  []byte
+		nextError   error
+
+		separatorToken []byte
+		getFn          GetFn
+		minPage        int
+		maxPage        int
+	}
+)
+
+// NewIterator constructs a new iterator.
+func NewIterator(
+	minPage int,
+	maxPage int,
+	getFn GetFn,
+	separatorToken []byte,
+) Iterator {
+	separatorTokenCopy := make([]byte, len(separatorToken), len(separatorToken))
+	copy(separatorTokenCopy, separatorToken)
+	itr := &iterator{
+		currentPage: -1,
+
+		separatorToken: separatorTokenCopy,
+		getFn:          getFn,
+		minPage:        minPage,
+		maxPage:        maxPage,
+	}
+	itr.advance(true)
```

Comment:
maybe just itr.init() instead of advance()?

Reply:
Its also called in next method though.